File: gcp_solver.py
import numpy as np
import networkx as nx
import matplotlib.pyplot as plt
import random
from random import randrange


class GCP_Solver:

    def __init__(self, graph, k, render_mode):

        self.graph = graph 
        self.k = k
        
        N = graph.order()
        colors = list(range(k))
        
        # Initialize solution with random colors
        self.solution = np.random.randint(0,self.k, size=N, dtype=np.int32)
        

        assert render_mode is not None, "render mode should be designated"
        self.render_mode = render_mode

        self.color_map = None
        self.layout = nx.spring_layout(self.graph,k=0.1)
        self.fig, self.ax = plt.subplots(figsize=(10,10))
        plt.ion()
    


    def render(self,solution):

        def is_in_ipython():
            try:
                __IPYTHON__
                return True
            except NameError:
                return False

        if is_in_ipython():
            from IPython.display import clear_output
            # clear_output(wait=True)


        if self.color_map is None:
            self.color_map = {
                    j:f"#{''.join([random.choice('0123456789ABCDEF') for i in range(6)])}" for j in range(self.k)
            }

        if self.layout is None:
            self.layout = nx.spring_layout(self.graph)


        #fig, ax = plt.subplots(figsize=(10, 10))
        #fig.tight_layout()
        #ax.axis("off")

        node_colors = [self.color_map[node] for node in solution]

        edge_colors = [
            "#ff0000" if solution[x] == solution[y] else "#000000"
            for x, y in nx.edges(self.graph)
        ]
        alphas = [
            1.0 if solution[x] == solution[y] else 0.1
            for x, y in nx.edges(self.graph)
        ]

        self.ax.clear()
        self.ax.axis('off')

        #plt.cla()
        nx.draw_networkx_nodes(self.graph, self.layout, node_color=node_colors)
        nx.draw_networkx_labels(self.graph, self.layout)
        nx.draw_networkx_edges(
            self.graph, self.layout, edge_color=edge_colors, alpha=alphas
        )


        # Display these info as text on the plot
        n_conflicts = self.count_conflicts(self.graph, solution)
        n_colors_used = len(np.unique(self.solution))

        info_text = f"Graph order: {self.graph.order()}\nGraph size: {self.graph.size()} \n\nNumber of Colors: {n_colors_used} \nConflicts: {n_conflicts}"
        self.ax.text(0.01, 0.99, info_text, transform=self.ax.transAxes, fontsize=10,
                     verticalalignment='top', bbox=dict(boxstyle='round', facecolor='wheat', alpha=0.5))

        self.fig.canvas.draw()
        self.fig.canvas.flush_events()
        plt.pause(0.1)


        if self.render_mode == "file":

            image_from_plot = np.frombuffer(self.fig.canvas.tostring_rgb(), dtype=np.uint8)
            image_from_plot = image_from_plot.reshape(
                self.fig.canvas.get_width_height()[::-1] + (3,)
            )

            img = Image.fromarray(image_from_plot)
            print("saving", f"{self.base_filename}{self.render_iter}.png")
            img.save(f"{self.base_filename}{self.render_iter}.png")
            self.render_iter += 1
            plt.close()

        elif self.render_mode == "human" or self.render_mode == "rgb_array":
            plt.show()

            # if capture_video, should return pixel array
            if self.render_mode == "rgb_array":
                self.fig.canvas.draw()
                image_from_plot = np.frombuffer(self.fig.canvas.tostring_rgb(), dtype=np.uint8)
                image_from_plot = image_from_plot.reshape(self.fig.canvas.get_width_height()[::-1] + (3,))
                return image_from_plot



    def tabucol(self,graph,k,tabu_size=7,reps=100,max_iterations=1000, alpha=0.6):

        """
        function tabucol() implements tabucol algorithm

            parameter

                graph: networkx graph
                k : number of colors using
                tabu_size: size of tabu list
                reps: number of iterations searching neighbor solution
                max_iterations: maximum iterations of tabucol
                alpha: parameter of duration equation

            return

                solution: solution list applied tabucol
        """
        N = graph.order()
        colors = list(range(k))

        # Tabu List
        tabu = {}
        iterations = 0

        print(f"Initial Solution: {self.solution}")
        
        conflicts = 0
        aspiration_level = dict()
        while iterations < max_iterations:

            conflicts = self.count_conflicts(graph, self.solution)
            conflicting_nodes = set()
            for node in graph.nodes():
                for neighbor in graph.neighbors(node):
                    if self.solution[node] == self.solution[neighbor]:
                        conflicting_nodes.add(node)
                        conflicting_nodes.add(neighbor)

            conflicting_nodes = list(conflicting_nodes)
            # Exit Condition
            if conflicts == 0:
                break


            # Exploring Neighbor solutions
            new_solution = None
            found_better = False
            for r in range(reps):
                node = conflicting_nodes[randrange(0,len(conflicting_nodes))]

                # allocate best color: generating minimum conflicts
                new_color = self.choose_color(graph, self.solution, node, colors)
                new_solution = self.solution.copy()
                new_solution[node] = new_color

                new_conflicts = self.count_conflicts(graph, new_solution)

                L = randrange(0,9)
                # duration = L + f(s)*alpha
                duration = int(L + new_conflicts * alpha)

                # If found a better solution,
                if new_conflicts < conflicts:
                    found_better = True

                    # if f(s') <= A(f(s))
                    if new_conflicts <= aspiration_level.setdefault(conflicts, conflicts-1):
                        # set A(f(s)) = f(s')-1
                        aspiration_level[conflicts] = new_conflicts - 1

                        if (node, new_color) in tabu and tabu[(node, new_color)] > iterations:
                            tabu.pop((node, new_color))
                        else:
                            tabu[(node, new_color)] = iterations + duration

                    # case for found good solution but that action is in tabu, then should search more
                    else:
                        if (node, new_color) in tabu:
                            continue
                        else:
                            tabu[(node, new_color)] = iterations + duration

                    break


            if found_better:
                self.solution = new_solution

            # Clean up expired tabu entries
            tabu = {move: expiry for move, expiry in tabu.items() if expiry > iterations}

            iterations += 1
            #if iterations % 500 == 0:
            print(f"Iteration: {iterations}, Conflicts: {conflicts}")
            print(f"Solution: {self.solution}")
            if iterations % 10 == 0:
                self.render(self.solution)

        print("final solution: ",self.solution)
        # After all iterations, return solution
        return self.solution



    def count_conflicts(self,graph, solution, node=None):
        """
        function count_conflicts() calculates number of conflicts given graph and solution.

            parameter

                graph: networkx graph

                solution:

                node: if node is given, only search conflicts around the node


            return

                conflicts: number of conflicts

        """

        conflicts = 0
        if node is None:
            for node in graph.nodes():
                for neighbor in graph.neighbors(node):
                    if solution[node] == solution[neighbor]:
                        conflicts += 1
            
            conflicts = conflicts // 2

        else:
            for neighbor in graph.neighbors(node):
                if solution[node] == solution[neighbor]:
                    conflicts += 1

        return conflicts




    def choose_color(self,graph, solution, node, colors):
        """
        function choose_color() chooses best color given target node.
        Best color means the color which makes minimal conflicts

            parameter

                graph: networkx graph

                solution:

                node: target node

                colors: color array


            return

                best_color

        """

        min_conflicts = float('inf')
        best_color = solution[node]
        for color in colors:
            if color != solution[node]:
                temp_solution = solution.copy()
                temp_solution[node] = color
                conflicts = self.count_conflicts(graph, temp_solution,node)
                if conflicts < min_conflicts:
                    min_conflicts = conflicts
                    best_color = color

        return best_color




    def solve(self):
        """
        function solve literally do tabucol algorithm

        """

        self.tabucol(self.graph, self.k)

File: gcp_solver.cpp

#include "gcp_solver.h"


GCP_Solver::GCP_Solver(const vector<vector<int>>& target_graph, int k, string algorithm)
: k(k), solution(target_graph.size(), 0), algorithm(algorithm){
    this->nodes = target_graph.size();
    //printf("constructor Node: %d\n",this->nodes);

    this->graph.resize(nodes, vector<int>(nodes, 0));
    this->graph = target_graph; 

    for(int i=0;i<nodes;i++)
	this->solution[i] = rand() % k;

    std::srand(std::time(nullptr)); // initialize random seed
}

void GCP_Solver::print_graph(){


    printf("print graph started\n");
    for(int u=0;u<this->nodes;u++){
        for(int v=0;v<this->nodes;v++)
            printf("%d ",this->graph[u][v]);

        printf("\n");
    }
}


void GCP_Solver::print_solution(const vector<int>& solution){

    for(int n=0;n<this->nodes;n++)
 	printf("%d ",solution[n]);
    printf("\n");

}



std::pair<vector<int>,int> GCP_Solver::solve(){

    int reward=0;
    if (algorithm == "tabucol")
	//printf("tabucol starting..\n");
    	return tabucol(this->solution, k);

    return {this->solution, reward};
}


std::pair<vector<int>,int> GCP_Solver::tabucol(const vector<int>& solution, int k, int tabu_size, int reps, int max_iterations, float alpha){

    /*
        function tabucol() implements tabucol algorithm

            parameter

                solution: initial solution
                k : number of colors using
                tabu_size: size of tabu list
                reps: number of iterations searching neighbor solution
                max_iterations: maximum iterations of tabucol
                alpha: parameter of duration equation

            return

                solution: final solution applied tabucol
    */
    int initial_conflicts = count_conflicts(solution);
    printf("Iterations: 0, Conflicts: %d\n",initial_conflicts);
    	
    vector<int> result_solution(nodes,0);
    result_solution = solution;

    // For Random Values
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> nodeDis(0, nodes-1);
    uniform_int_distribution<> randDis(0, 8);    


    int iterations = 0;
    int conflicts = 0;
    unordered_map<int, int> aspiration_level;

    while (iterations < max_iterations){
    
	conflicts = count_conflicts(result_solution);
	unordered_set<int> conflicting_nodes;
	for (int u=0; u<nodes; u++){
	    for (int v=0; v<nodes; v++){
		if(graph[u][v] && (result_solution[u] == result_solution[v])){
		   conflicting_nodes.insert(u);
		   conflicting_nodes.insert(v);
		}
            }

	}

	// Exit Condition
	if (conflicts == 0)
           break;

	vector<int> conflicting_nodes_list(conflicting_nodes.begin(), conflicting_nodes.end());
    	uniform_int_distribution<> conf_nodeDis(0, conflicting_nodes_list.size()-1);
	vector<int> new_solution(nodes,0);

	for (int r=0; r < reps; r++){

	   int node = conflicting_nodes_list[conf_nodeDis(gen)];

	   int new_color = choose_color(result_solution, node);
	   new_solution = result_solution;
	   new_solution[node] = new_color;

	   int new_conflicts = count_conflicts(new_solution);
	   
	   int L = randDis(gen);
	   int duration = int(L + new_conflicts * alpha);

	   if (new_conflicts < conflicts){
	      
	      if (new_conflicts <= aspiration_level[conflicts]){
		  aspiration_level[conflicts] = new_conflicts - 1;
		  auto tabu_entry = tabu_list.find({node, new_color});
		  if (tabu_entry != tabu_list.end() && tabu_entry->second > iterations)
		      tabu_list.erase(tabu_entry);
		  else
	  	      tabu_list[{node,new_color}] = iterations + duration;
	      
              // case for found good solution but that action is in tabu, then should search more
	      }else{
		  
		  if (tabu_list.find({node, new_color}) != tabu_list.end())
		      continue;
		  else
 		      tabu_list[{node, new_color}] = iterations + duration;
              }

	      break;
	   }

    	}  
  
	result_solution = new_solution; 

 	iterations += 1;
    	if (iterations % 500 == 0){
            printf("Iterations: %d, Conflicts: %d\n",iterations, conflicts);
    	}	

	// Clean up expired tabu entries
        for (auto it = tabu_list.begin(); it != tabu_list.end();) {
            if (it->second <= iterations) {
                it = tabu_list.erase(it);
            } else {
                ++it;
            }
        }
    }

    
    // After while loop, calculate reward

    int final_conflicts = count_conflicts(result_solution);
    
    int tabucol_reward = final_conflicts - initial_conflicts;
	
    //printf("final conflict: %d\n", final_conflicts);    

    // After all iterations, return reward
    return {result_solution, tabucol_reward};

}



int GCP_Solver::count_conflicts(const vector<int>& solution, int node){
    /*
        function count_conflicts() calculates number of conflicts given graph and solution.

            parameter

                graph: networkx graph

                solution:

                node: if node is given, only search conflicts around the node


            return

                conflicts: number of conflicts

    */
	
    int conflicts = 0;
    if(node == -1){
        for (int u=0;u<this->nodes;u++){
	    for (int v = u+1; v<this->nodes;v++){
		  if(this->graph[u][v] && (solution[u] == solution[v])){
			conflicts += 1;
		  }
	    }
	}
	
	conflicts /= 2;

    }else{

	for (int v=0; v<this->nodes;v++){
	    if (this->graph[node][v] && (solution[node] == solution[v])){
		  conflicts += 1;
	    }
	}
    }
    

    return conflicts;
}




int GCP_Solver::choose_color(const vector<int>& solution, int node, float p){

    /*
        function choose_color() chooses best color given target node.
        Best color means the color which makes minimal conflicts

            parameter

                graph: networkx graph

                solution:

                node: target node

                colors: color array


            return

                best_color

    */

    int min_conflicts = MAX_NUM;
    int best_color = solution[node];
    
    vector<int> colors(k);
    iota(colors.begin(), colors.end(), 0);    
    
    double rand_prob = static_cast<double>(std::rand()) / RAND_MAX;
    bool choose_random = (rand_prob > p);

    if (choose_random){
      vector<int> available_colors;
      for (int color : colors){
 	if (color != solution[node])
	   available_colors.push_back(color);
      }
	
      if (!available_colors.empty()){
	  int rand_idx = std::rand() % available_colors.size();
	  return available_colors[rand_idx];
      } else
	  return best_color;


    } else {
	// choose most promising color (that decreases conflicts most)
	
    	for (int color : colors){
	   if (color != solution[node]){
	   	vector<int> temp_solution = solution;
	   	temp_solution[node] = color;
	   	int conflicts = count_conflicts(temp_solution, node);
	   	if (conflicts < min_conflicts){
		    min_conflicts = conflicts;
		    best_color = color;
	   	}
           }

        }

    }

    return best_color;


}



